### ISSUE TITLE
proxies for recursive concise methods have the wrong `.length` property

### ISSUE THREAD

**Issue by raganwald** [babel/babel#896]
```javascript
const foo = {
  fact (n) {
    return n < 3
      ? n
      : fact(n-1) * n
  },
  bar (n) {
    return this.fact(n);
  }
};

foo.fact.length
  //=> 0
foo.bar.length
  //=> 1
```

Shouldn’t `foo.fact.length` and `foo.bar.length` both evaluate to `1`?

**Comment by thejameskyle** [comment]
That code isn't correct to begin with, you mean to reference `this.fact()` within `foo.fact` which gives it the correct length. 

```js
const foo = {
  fact (n) {
    return n < 3
      ? n
      : this.fact(n-1) * n
  },
  bar (n) {
    return this.fact(n);
  }
};

console.log(foo.fact.length)
  //=> 1
console.log(foo.bar.length)
  //=> 1
```

**Comment by raganwald** [comment]
No, I left it that way for a purpose, as it triggers the compiler to treat `fact` as if it were a named function. If you write `this.fact`, it is no longer invoking the name of the function recursively.

**Comment by getify** [comment]
@raganwald yes, but as written, that code will break on invoking either `foo.fact(..)` or `foo.bar(..)` because of lack of a `fact` lexical reference on line 5 of the OP snippet.

**Comment by raganwald** [comment]
That is a red herring. Not only does it not break, but this displays the exact same behaviour:

```javascript
function fact (n) {
  return n < 3 ? n : n * fact(n-1);
}

const foo = {
  fact (n) {
    return n < 3
      ? n
      : fact(n-1) * n
  },
  bar (n) {
    return this.fact(n);
  }
};

console.log(foo.fact.length)
  //=> 0
```

**Comment by sebmck** [comment]
Fixed, thanks!

**Comment by raganwald** [comment]
Thank you so much! :gift:

**Comment by sebmck** [comment]
Fixed as of 4.5.2, thanks again!
